sql-plugin/src/main/scala/org/apache/spark/sql/rapids/execution/GpuHashJoin.scala

```diff
@@ -267,8 +267,12 @@ abstract class BaseHashJoinIterator(
   // We can cache this because the build side is not changing
   private lazy val streamMagnificationFactor = joinType match {
     case _: InnerLike | LeftOuter | RightOuter | FullOuter =>
-      withResource(GpuProjectExec.project(built.getBatch, boundBuiltKeys)) { builtKeys =>
-        guessStreamMagnificationFactor(builtKeys)
+      built.allowSpilling()
```

Comment:
I think you need to use `build.checkpoint()` here, and then use a `withRestoreOnRetry` below to ensure we close the cached `ColumnarBatch` in `built` in the case of a retry.